### ISSUE TITLE
Wrong equals operation for points leads to wrong save behavior in EntityFramework

### ISSUE THREAD

**Issue by Danielku15** [NetTopologySuite/NetTopologySuite#447]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of NetTopologySuite
- [x] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/NetTopologySuite/NetTopologySuite/issues) and [closed](https://github.com/NetTopologySuite/NetTopologySuite/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
NTS has a design issue that the equality check of 2 points is only done on 2D base. We are having the issue that we are updating some coordinates to add the height component if it is missing. But due to the fact that `new Point(1,2) == new Point(1,2,3)` (including `Equals` and `EqualsExact`), Entity Framework will not detect any change in the value and will not generate an update statement.

![image](https://user-images.githubusercontent.com/674916/98841901-0d57c300-2449-11eb-9fca-cdb7a215545b.png)

### Steps to Reproduce

```csharp
Assert.IsFalse(new NetTopologySuite.Geometries.Point(1,2) == new NetTopologySuite.Geometries.Point(1,2,3))
Assert.AreNotEqual(new NetTopologySuite.Geometries.Point(1,2), new NetTopologySuite.Geometries.Point(1,2,3))
```

### System Configuration
Tell us about the environment where you are experiencing the bug:
- NetTopologySuite version: 2.0.0
- Other NetTopologySuite packages and versions: 
  - Microsoft.EntityFrameworkCore.SqlServer.NetTopologySuite 3.1.6
  - NetTopologySuite.IO.SqlServerBytes 2.0.0
- Environment (Operating system, version and so on): Windows 10 Enterprise 2004
- .NET Framework version: .net 4.8
- Additional information: 

Here we can see that the equals operation is only respecting 2D by design: 

https://github.com/NetTopologySuite/NetTopologySuite/blob/7babd53a14f452ec12aee6c15a6f5f943d26aa5d/src/NetTopologySuite/Geometries/Coordinate.cs#L259-L262

NTS should correctly respect the availability of the height component and consider it for equality checks. Especially in regards to EntityFramework where NTS is also used for Spatial Operations it is very essential to respect also the height component. Even more when working in the geodetic domain.

**Comment by DGuidi** [comment]
actually this is a design decision from JTS (so NTS inherits this), because algorithms are planar. change this behaviour basically break a lot of things , I presume (I remember the worst days slon my life spent finding bugs year ago before discover that I override the equality operators without actually knowing what I'm doing...)

**Comment by Danielku15** [comment]
I might be opening here some fundamental discussion but to me NTS is not really suitable for real world GIS or GNSS applications if things are only 2D. I can understand that NTS might focus to 2D geometry (as the readme indicates), and this originates from JTS. But NTS grew already way beyond that. The first sentence in the readme mentions "A .NET GIS solution that is fast and reliable for the .NET platform". I work at a market leader for GNSS and GIS products, if I tell our people that height components are not relevant in GIS or GNSS they would just look at me confused what I'm talking about. Having a pipe 1,2, or 5meters below the street is very important. Telling a drone that it's geographical position is 2 meters or 10 meters above the earth surface is very important for steering.

NTS is the main library to be used for Spatial and Geometric types in the Entity Framework of Microsoft, one primary ORM if you're developing in the .net world. There is also not really a simple alternative that as a developer/company could say if NTS is 2D only we use something else that provides 3D capabilities.

**Comment by FObermaier** [comment]
A proposal:
* Add static `CoordinateComparer` to Geometry class
```C#
public static CoordinateEqualityComparer CoordinateComparer { get; set; }` 
```
* `CoordinateEqualityComparer` class definition
```C#
/// <summary>
/// A class that can be used to compare coordinates for equality
/// </summary>
public abstract class CoordinateEqualityComparer
{
    /// <summary>
    /// Method to test 2 <see cref="Coordinate"/>s for equality, allowing a tolerance.
    /// </summary>
    /// <param name="a">The 1st Coordinate</param>
    /// <param name="b">The 2nd Coordinate</param>
    /// <param name="tolerance">A tolerance value</param>
    /// <returns></returns>
    public abstract bool AreEqual(Coordinate a, Coordinate b, double tolerance);
}
```
* Possible implementation for CoordinateZ
```C#
public sealed class CoordinateZComparer : CoordinateEqualityComparer
{
    public override bool AreEqual(Coordinate a, Coordinate b, double tolerance)
    {
        if (tolerance.Equals(0d))
            return a.Equals(b) && EqualInZ(a, b);

        return a.Distance(b) <= tolerance && DistanceInZ(a, b) <= tolerance;
    }

    private static bool EqualInZ(Coordinate a, Coordinate b)
    {
        if (double.IsNaN(a.Z) && double.IsNaN(b.Z))
            return true;

        if (double.IsNaN(a.Z) || double.IsNaN(b.Z))
            return false;

        return a.Z == b.Z;
    }

    private static double DistanceInZ(Coordinate a, Coordinate b)
    {
        if (double.IsNaN(a.Z) && double.IsNaN(b.Z))
            return 0d;

        if (double.IsNaN(a.Z) || double.IsNaN(b.Z))
            return double.PositiveInfinity;

        return Math.Abs(a.Z - b.Z);
    }
}
```
* Wire all this in `protected static bool Geometry.Equal(Coordinate, Coordinate, double)`
```C#
protected static bool Equal(Coordinate a, Coordinate b, double tolerance)
{
    var cc = CoordinateComparer;
    if (cc != null) return cc.AreEqual(a, b, tolerance);

    if (tolerance == 0)
        return a.Equals(b);

    return a.Distance(b) <= tolerance;
}
```

Assigning `CoordinateZComparer` as default, **ALL but 1** NUnit tests are looking good. The new failure case is `NetTopologySuite.Tests.NUnit.Triangulate.DelaunayTest.TestTriangle`, a test where 3d-input is compared to 2d-output.

@bricelam, @airbreather: Are there any options in EF to solve the initial issue, other suggestions?

**Comment by airbreather** [comment]
locationtech/jts#366 touches this kind of issue.

> I think this is quite a critical topic which needs to be addressed within NTS. I fully understand that such a change needs proper design and consideration. Existing applications need to continue like before, but there should be an option which allows developers to tell NTS to work in 3D. e.g. introduce a Point3D, a setting, additional methods, etc.

I agree that there's a bit of a gap in our toolset when it comes to 3D.  However, I have a concern that the common standard for this kind of stuff ([OGC Simple Feature Access](https://www.ogc.org/standards/sfa)) says quite clearly in §6.1.2.5:

> Spatial operations work in the "map geometry" of the data and will therefore not reflect z or m values in calculations (e.g., Equals, Length) or in generation of new geometry values (e.g., Buffer, ConvexHull, Intersection). This is done by projecting the geometric objects onto the horizontal plane to obtain a "footprint" or "shadow" of the objects for the purposed of map calculations. In other words, it is possible to store and obtain z (and m) coordinate values but they are ignored in _all_ other operations which are based on map geometries. Implementations are free to include true 3D geometric operations, but should be consistent with ISO 19107.

The underlying problem here is really an extension of the same issues from my "Unproductive Rant" [back over here](https://github.com/NetTopologySuite/NetTopologySuite/issues/233#issuecomment-455780259): our model is significantly out-of-sync with the inconvenient details of reality, and the only reason why we can get away with it is because the math in this model is much easier to implement much more efficiently than a more accurate and complete model.  *"Remember that all models are wrong; the practical question is how wrong do they have to be to not be useful." -George E. P. Box*

A more accurate and complete model for terrestrial calculations should be able to naturally extend to things like "a pipe that's 5 meters above the ground" and be able to make interesting calculations in 3D space based on those concepts.

I know of no existing standards out there that speak to this, however.  Such a model would need to be able to talk about not just Point / LineString / Polygon / MultiX, but also things like Spheres / Cylinders / arbitrary figures in space.  That doesn't mean that there **aren't** any standards out there (perhaps I just don't know about them), nor that there **must** be a standard in order for work to begin on this, but it does limit the reach of such a solution and puts a lot of burden onto developers who aim to build that solution.

If we're intending to deviate from JTS and from the standard on this particular issue, I would look at GEOS and, perhaps, other JTS derivatives to see if (and how?) they have tackled this issue.

-----

> other suggestions?

For equality, specifically, I think it's not a bad idea to explore deviating from JTS.  My immediate, top-of-my-head, never-before-compiled, pre-morning-coffee suggestion for changing equality testing would look something like this, to ensure that coordinates are never considered "equal" if they disagree on which dimensions are carried by the Coordinate instance:
```csharp
protected static bool Equal(Coordinate a, Coordinate b, double tolerance)
{
    int dimensionA = Coordinates.Dimension(a);
    int dimensionB = Coordinates.Dimension(b);
    if (dimensionA != dimensionB)
    {
        return false;
    }

    int measuresA = Coordinates.Measures(a);
    int measuresB = Coordinates.Measures(b);
    if (measuresA != measuresB)
    {
        return false;
    }

    for (int dim = 0; dim < dimensionA; dim++)
    {
        double valA = a[dim];
        double valB = b[dim];

        // this next line is carefully crafted to ensure reasonable treatment of NaNs
        if (!(valA.Equals(valB) || Math.Abs(valA - valB) <= tolerance))
        {
            return false;
        }
    }

    return true;
}
```

With this in place, the only backwards-compat we should need is an "all-or-nothing" flag to opt-in to the new behavior.

**Comment by rhuijben** [comment]
While I agree that this is problematic on the Geometry side we really break the .Net rules for .Equals() here, and that is causing many other problems.  Geometric compares shouldn't use the ordinary .Equals(), .GetHashCode(), IEquatable<T>, etc as we can't just make the rest of .Net handle things as equal that are not.

Fixing this would also allow using Lynq in quite a few NTS usages where it now falls apart on .Equals() not being compatible.


I think we should first be a good .Net citizen and then allow different usecases. Some of the behavior settings could be delegated to the GeometryFactory (like how we handle SRID), but putting different points in an hashtable, hashset, etc. should just work.

**Comment by FObermaier** [comment]
@airbreather, `Geometries.Coordinates.Dimension(Coordinate)` returns how many ordinates a `Coordinate` can carry, not how many meaningful values it has (`Coordinate.NullOrdinate`). Likewise `Geometries.Coordinates.Measures` answers how many measure values a `Coordinate` can carry. 

This causes the following test not to pass
```C#
var pt1 = new NetTopologySuite.Geometries.Point(new NetTopologySuite.Geometries.Coordinate(10, 10));
var pt2 = new NetTopologySuite.Geometries.Point(new NetTopologySuite.Geometries.CoordinateZ(10, 10));
Assert.IsTrue(pt1 == pt2); 
```

**Comment by Danielku15** [comment]
I like seeing so many people participating on this discussion, it really shows the complexity and the challenge behind it. I personally like the statement of @rhuijben that the main issue might come from the fact that the .net rules for certain operations are not followed, at least for this particular case on equality. 

Of course you guys have way more insight on all the internals of JTS and NTS conceptually and architecturally so my following remarks are more a "brainstorming" than a well founded proposal 😉 There are many solutions to this which have all pros and cons. 

If I would personally design such a library like NTS, I would definitly rework and change the concept many time before it really fits well. In the past for me the best thing that worked out is to keep data and logic separated. In NTS operations and data are tightly coupled within the corresponding classes. If I would start from scratch using latest tech, I would likely put all data classes into C# records. Points, Lines,... This way the rules of .net are followed regarding equality, hashcode etc. For the operations either extension methods or facades would provide the logic: 

- You need 2D distance operations according to ISO 19125? -> Use the `Iso19125GeometryOperations.Distance(point, point)` 
- You need geospatial operations respecting 3D information on a WGS84 system? Use the `Wgs84SpatialOperations.Distance(point, point)`. 
- You have an own 3rd party library doing this job respecting all kind of SRIDs? We have you covered with a `Proj4NetOperations.Distance(point, point, /*whatever proj4net needs as input */)` in a `NetTopologySuite.Proj4Net` NuGet package. 

Benefit #1: Data classes are simplistic and .net compliant. No "assumptions" or "algorithmic aspects" influence them. Good old POCOs which can be used in many different ways 😁  
Benefit #2: Devs are enforced to "choose" the "right" operation provider for their needs.

As @airbreather quoted so nice: "Remember that all models are wrong". At least such a system let's devs choose which 'wrong' is 'good' for their needs. 


Phew. 😅💦 I hope this was not too off-topic as such a change will not just be possible by tomorrow. As an intermediate step I would say moving at least the equality operations of all classes to a place where custom `IEqualityComparer` instances can be set will solve my issues. I could just set my custom comparer there and get my 3D comparisons applied also on EF core level when queries are generated. Having new overloads or other will not help as the `Equals()` call is happening within the EF core change tracker.

**Comment by bricelam** [comment]
See [overriding default value comparers](https://docs.microsoft.com/en-us/ef/core/modeling/value-comparers#overriding-defaults) in the EF Core docs. I'm not sure if it's possible with Geometry properties, but that's where I'd start when looking for a workaround.

**Comment by FObermaier** [comment]
This issue has now been addressed. We are now able to specify a `CoordinateEqualityComparer` class that carries out the comparison. 2 implementations are provided:
* `CoordinateEqualityComparer` is the default and mimics the implementation of NTS up to version v2.1
* `PerOrdinateEqualityComparer` compares each ordinate of two coordinates to be equal within a given tolerance.

To use it, you need to set up a `NtsGeometryServices` instance and use that to create `GeometryFactory`s and use those to create your geometries.

For an example see:
https://github.com/NetTopologySuite/NetTopologySuite/blob/develop/test/NetTopologySuite.Samples.Console/Tests/Github/Issue447.cs

**Comment by Danielku15** [comment]
Looks like a great solution. Well designed and flexible. But unfortunately it only solves the issue if you have all object creations in your hand. 

I guess we will need to reach out to the EFCore team to start using this factory feature otherwise the equality checks within the EF change tracker will still fail and not trigger updates (starting point of this topic). If EFCore creates the objects using a different factory when materializing you end up with a mix of points which might use different comparers. In such cases flipping operators would lead to different results: 

```myPointFromFactory.EqualsExact(efPoint) == true`` (because custom comprarer is used)
```efPoint.EqualsExact(myPointFromFactory) == false`` (because default comprarer is used)

**Comment by FObermaier** [comment]
You should be all set if you use
```
NtsGeometryServices.Instance = <your configured NtsGeometryServices>;
```

**Comment by Danielku15** [comment]
I just looked at the latest EFCore source state and they are using already the NtsGeometryServices: 
https://github.com/dotnet/efcore/blob/1d9589691029d1a1b33aeb4e90eb846cddd4382f/src/EFCore.SqlServer.NTS/Extensions/SqlServerNetTopologySuiteServiceCollectionExtensions.cs#L31

From there they use the [SqlServerBytesReader](https://github.com/NetTopologySuite/NetTopologySuite.IO.SqlServerBytes/blob/ff02ec98634ac2dcc0c1cffa53771f5c24d96cd9/src/NetTopologySuite.IO.SqlServerBytes/SqlServerBytesReader.cs) which should complete the loop. 

I was concerned that EFCodre simply creates new instances `new Point()` without utilizing the factory. 

So my concerns were not founded well 😄 Looking forward to be able to upgrade and remove our local hacks for comparison.

**Comment by FObermaier** [comment]
> Of course I could build my own comparer and activate it, but I would guess it makes sense for all users of NTS assuming geographical usage.

You could contribute one with a PR!

The current `PerOrdinateEqualityComparer` only uses a tolerance value when `Geometry.EqualsExact(Geometry geom, double tolerance)` is called. I think you want a comparer that chips in tolerance values even when `Geometry.EqualsExact(Geometry)` is called.

**Comment by FObermaier** [comment]
I assume EF.Core relys on `Geometry.Equals(object|Geometry)` which in turn calls `Geometry.EqualsExact(Geometry)` without the tolerance value, so there is probably no way to chip in a tolerance `Coordinate` here.

I think the tolerance `Coordinate` would have to be an instance member of your `CoordinateEqualityComparer` implementation.
You could then override `AreEqual` like this:
```C#
// Be sure that _toleranceCoord.Z and _toleranceCoord.M are not double.NaN!
private readonly CoordinateZM _toleranceCoord;

protected override bool AreEqual(Coordinate a, Coordinate b, double tolerance)
{
    return Distance(a.X, b.X) <= Math.Min(_toleranceCoord.X, tolerance) &&
           Distance(a.Y, b.Y) <= Math.Min(_toleranceCoord.Y, tolerance) &&
           Distance(a.Z, b.Z) <= Math.Min(_toleranceCoord.Z, tolerance) &&
           Distance(a.M, b.M) <= Math.Min(_toleranceCoord.M, tolerance);
}
```

**Comment by FObermaier** [comment]
Fixed with 8d94d5e8eb4d6bb5a360e15be74d9e0ae63ecc46
